modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxFinishFuture.java

```diff
@@ -978,6 +988,19 @@ public GridDistributedTxMapping mapping() {
 
         /** {@inheritDoc} */
         @Override boolean onNodeLeft(UUID nodeId, boolean discoThread) {
+            if (tx.state() == COMMITTING || tx.state() == COMMITTED) {
+                if (concat(of(m.primary().id()), Optional.of(tx.transactionNodes().getOrDefault(nodeId, Collections.emptySet())).orElse(Collections.emptySet()).stream())
```

Comment:
Weird code, must be:
Optional.of(tx.transactionNodes().getOrDefault(nodeId, Collections.emptySet())).orElse(Collections.emptySet()).stream()) -> tx.transactionNodes().getOrDefault(nodeId, emptySet()).stream()

Never forget about static imports.